tools/cldr-code/src/main/java/org/unicode/cldr/util/personname/PersonNameFormatter.java

```diff
@@ -0,0 +1,524 @@
+            ModifiedField that = (ModifiedField) obj;
+            return field == that.field && modifiers.equals(that.modifiers);
+        }
+        @Override
+        public int hashCode() {
+            return field.hashCode() ^ modifiers.hashCode();
+        }
+    }
+
+    /**
+     * An element of a name pattern: either a literal string (like ", ") or a modified field (like {given-initial})
+     * The literal is null IFF then modifiedField is not null
+     * Immutable
+     */
+    public static class NamePatternElement {
+        private final String literal;
+        private final ModifiedField modifiedField;
+
+
+        public String getLiteral() {
+            return literal;
+        }
+        public ModifiedField getModifiedField() {
+            return modifiedField;
+        }
+
+        /**
+         * @param literal
+         * @param field
+         * @param modifiers
+         */
+        public NamePatternElement(ModifiedField modifiedField) {
+            this.literal = null;
+            this.modifiedField = modifiedField;
+        }
+
+        public NamePatternElement(String literal) {
+            this.literal = literal;
+            this.modifiedField = null;
+        }
+
+        /** convenience method for testing */
+        public static NamePatternElement from(Object element) {
+            if (element instanceof ModifiedField) {
+                return new NamePatternElement((ModifiedField) element);
+            } else {
+                String string = element.toString();
+                if (string.startsWith("{") && string.endsWith("}")) {
+                    return new NamePatternElement(ModifiedField.from(string.substring(1, string.length()-1)));
+                } else {
+                    return new NamePatternElement(string);
+                }
+            }
+        }
+        @Override
+        public String toString() {
+            // TODO: escape \ and { in literals
+            return literal != null ? literal : modifiedField.toString();
+        }
+    }
+
+    /**
+     * A name pattern, corresponding to a string such as "{given-initial} {surname}"
+     * Immutable
+     */
+    public static class NamePattern {
+        private final List<NamePatternElement> elements;
+        private final Set<Field> fields;
+
+        public Set<Field> getFields() {
+            return fields;
+        }
+
+        public String format(NameObject nameObject) {
+            StringBuilder result = new StringBuilder();
+            Set<Modifier> unhandledModifiers = EnumSet.noneOf(Modifier.class);
+
+            for (NamePatternElement element : elements) {
+                final String literal = element.getLiteral();
+                if (literal != null) {
+                    result.append(literal);
+                } else {
+                    final String bestValue = nameObject.getBestValue(element.getModifiedField(), unhandledModifiers);
+                    if (unhandledModifiers.isEmpty()) {
```

Comment:
This is likely `if (!unhandledModifiers.isEmpty())`

Reply:
fixed